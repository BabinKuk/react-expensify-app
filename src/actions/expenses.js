import uuid from 'uuid';
import database from '../firebase/firebase';

// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// with firebase and async redux actions
// component calls action generator
// action generator returns function
// component dispatches function
// function runs (has the ability to dispatch other functions amd do other things)

// Actions
// ADD_EXPENSE - changes redux store
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense: expense
});

// saving data to db - async action
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        // destructure data and set default
        const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;
        const expense = { description, note, amount, createdAt };

        // access firebase and save data
        return database.ref('expenses').push(expense).then((ref) => {
            // dispatch data
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

// REMOVE_EXPENSEc
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

// removing expense from db - async action
export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        // remove data
        return database.ref(`expenses/${id}`).remove().then(() => {
            // dispatch data
            dispatch(removeExpense({ id }));
        });
    };
};

// EDIT_EXPENSE - changes redux store
export const editExpense = ((id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
}));

// SET_EXPENSES - expenses array from db
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

// Fetch the data from db and dispatch - async action
export const startSetExpenses  = (expenseData = {}) => {
    return (dispatch) => {
        // 1. fetching data arrays once
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];
            console.log(snapshot.val());
            
            // 2. parse data into array
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            console.log(expenses);
            // 3. dispatch set_expenses
            dispatch(setExpenses(expenses));
        });
    };
};
