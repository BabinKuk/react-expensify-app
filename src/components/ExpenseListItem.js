// Export a stateless functional component
// description, amount, createdAt

import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import '../locales/hr';
import numeral from 'numeral';

numeral.locale('hr');

export const ExpenseListItem = ({ dispatch, id, description, amount, createdAt}) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>
            {numeral(amount / 100).format()} 
             - 
            {moment(createdAt).format('DD.MM.YYYY.')}
        </p>
    </div>
);

export default ExpenseListItem;
