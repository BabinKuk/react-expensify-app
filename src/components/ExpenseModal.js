import React from 'react'; 
import Modal from 'react-modal';

const ExpenseModal = (props) => (
    <Modal
        isOpen={!!props.deleteSelected}
        onRequestClose={props.handleClearSelectedExpense}
        contentLabel="Plese confirm to delete selected expense"
        closeTimeoutMS={300}
        className="modal"
    >
        <h3 className="modal__title">Please confirm to delete selected expense</h3>
        {props.deleteSelected && <p className="modal__body">{props.deleteSelected.description}</p>}
        <button className="button" onClick={props.onRemove}>Delete</button>
    </Modal>
);

export default ExpenseModal;
