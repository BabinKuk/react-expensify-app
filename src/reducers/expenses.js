// expenses reducer
const expensesReducerDefaultState = [];
const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            // spread array, add new expense
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            // remove from array
            return state.filter(({ id }) => {
                return id !== action.id; 
            });
        case 'EDIT_EXPENSE':
            // loop array and change the match
            return state.map((expense) => {
                if (expense.id === action.id) {
                    // change
                    // spread object with existing and new properties
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    // no change
                    return expense;
                }
            });
        default:
            return state;
    }
};

export default expensesReducer;