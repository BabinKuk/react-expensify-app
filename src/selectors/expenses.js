import moment from 'moment';

// Get visible expenses
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        // check if dates are valid
        //const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        //const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
        // check dates using moment.js
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;

        // figure out if description has the text variable string inside of it
        // using includes and tolowercase
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        
        //display if all three match
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            // recent date first
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === 'amount') {
            // greater amount first
            return a.amount < b.amount ? 1 : -1;
        }
    });
};


export { getVisibleExpenses };