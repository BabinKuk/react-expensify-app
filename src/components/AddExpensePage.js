import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        //console.log(expense);
        // sync with store
        //this.props.dispatch(addExpense(expense));
        // sync with store THIS WAY!!!!!
        this.props.startAddExpense(expense);
        // redirect page
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm 
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        );
    }
}

// abstract away dispatch from the component
const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);