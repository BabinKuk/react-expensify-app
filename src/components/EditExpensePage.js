import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        //console.log('editExpense onSubmit');
        // sync with store
        //this.props.dispatch(editExpense(this.props.expense.id, expense));
        // sync with store THIS WAY!!!!!
        this.props.startEditExpense(this.props.expense.id, expense);
        // redirect page
        this.props.history.push('/');
    };

    onRemove = () => {
        console.log('remove');
        // sync with store
        this.props.startRemoveExpense({ id: this.props.expense.id });
        // redirect page
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <h1>Edit Expense</h1>
                <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onRemove}>Remove</button>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => ({
    // get expense with given id
    expense: state.expenses.find((expense) => {
        return expense.id === props.match.params.id;
    })
});

// abstract away dispatch from the component
const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);