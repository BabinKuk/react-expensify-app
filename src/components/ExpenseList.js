import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses found</p>
            ) : (
                props.expenses.map((expense) => {
                    return <ExpenseListItem key={expense.id} {...expense} />
                })
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    // display selected data
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    };
};

// HOC with props from the store
export default connect(mapStateToProps)(ExpenseList);
