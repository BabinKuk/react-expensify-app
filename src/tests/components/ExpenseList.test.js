import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

//expense list takes in the array and renders new list item from the array
test('Test render ExpenseList with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});

test('Test render ExpenseList with empty messages array', () => {
    const wrapper = shallow(<ExpenseList expenses={[]}/>);
    expect(wrapper).toMatchSnapshot();
});