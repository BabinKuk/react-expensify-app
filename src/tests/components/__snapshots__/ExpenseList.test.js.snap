// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test render ExpenseList with empty messages array 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No expenses found
      </span>
    </div>
  </div>
</div>
`;

exports[`Test render ExpenseList with expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <ExpenseListItem
      amount={100}
      createdAt={0}
      description="voda"
      id="1"
      key="1"
      note=""
    />
    <ExpenseListItem
      amount={134500}
      createdAt={-345600000}
      description="plin"
      id="2"
      key="2"
      note=""
    />
    <ExpenseListItem
      amount={345100}
      createdAt={345600000}
      description="struja"
      id="3"
      key="3"
      note=""
    />
  </div>
</div>
`;
